README.md
# gilded-rose
Clojure implementation of the gilded rose kata, illustrating how simple the problem can be when you're not wasting time trying to figure out how to use design patterns, class wrappers, etc. as required in the original C# code.  

See https://github.com/emilybache/GildedRose-Refactoring-Kata for a collection of implementations across different languages.  Most do not implement a test suite as that is part of the exercise, this repo includes the tests I felt necessary to cover all the cases.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-07-18
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-07-18
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/gilded-rose/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/gilded-rose/compare/0.1.0...0.1.1


test/gilded_rose/gilded1_test.clj
(ns gilded-rose.gilded-test
  (:require [clojure.test :refer :all]
            [gilded-rose.gilded :refer :all]))

(deftest legendary
  (let [lgnd {:name "Sulfuras, Hand of Ragnaros" :sell-in 0 :quality 80}
        daily20 (take 20 (iterate update-quality lgnd))]
    (every? (partial = lgnd) daily20)))

(deftest conjured
  (let [c {:name "Conjured Mana Cake" :sell-in 5 :quality 15}]
    (are [c-static c-calc] (= c-static c-calc)
      {:name "Conjured Mana Cake" :sell-in 5 :quality 15} (days c 0)
      {:name "Conjured Mana Cake" :sell-in 0 :quality 5} (days c 5)
      {:name "Conjured Mana Cake" :sell-in -1 :quality 1} (days c 6)
      {:name "Conjured Mana Cake" :sell-in -3 :quality 0} (days c 8)
      {:name "Conjured Mana Cake" :sell-in -95 :quality 0} (days c 100))))

(deftest normal
  (let [e {:name "Elixir of the Mongoose" :sell-in 5 :quality 10}]
    (are [e-static e-calc] (= e-static e-calc)
      {:name "Elixir of the Mongoose" :sell-in 5 :quality 10} (days e 0)
      {:name "Elixir of the Mongoose" :sell-in 4 :quality 9} (days e 1)
      {:name "Elixir of the Mongoose" :sell-in 0 :quality 5} (days e 5)
      {:name "Elixir of the Mongoose" :sell-in -2 :quality 1} (days e 7)
      {:name "Elixir of the Mongoose" :sell-in -95 :quality 0} (days e 100))))

(deftest tickets
  (let [t {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in 15 :quality 20}]
    (are [t-static t-calc] (= t-static t-calc)
      {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in 15 :quality 20} (days t 0)
      {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in 14 :quality 21} (days t 1)
      {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in 10 :quality 25} (days t 5)
      {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in 7 :quality 31} (days t 8)
      {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in 3 :quality 41} (days t 12)
      {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in 1 :quality 47} (days t 14)
      {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in 0 :quality 50} (days t 15)
      {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in -85 :quality 0} (days t 100))))

(deftest aged
  (let [b {:name "Aged Brie" :sell-in 2 :quality 10}]
    (are [b-static b-calc] (= b-static b-calc)
      {:name "Aged Brie" :sell-in 2 :quality 10} (days b 0)
      {:name "Aged Brie" :sell-in 1 :quality 11} (days b 1)
      {:name "Aged Brie" :sell-in 0 :quality 12} (days b 2)
      {:name "Aged Brie" :sell-in -3 :quality 18} (days b 5)
      {:name "Aged Brie" :sell-in -98 :quality 50} (days b 100))))


src/gilded_rose/gilded1.clj
(ns gilded-rose.gilded
  (:gen-class))

(def items
  "Can't change this.  In the original OO code, list<items> was a property
  of the class which could not be changed, and each item, represented as
  a hash-map here, was an Item class, which also could not be changed.

  We will honor these constraints by not allowing the structure of items to
  be changed."

  [{:name "+5 Dexterity Vest" :sell-in 10 :quality 20}
   {:name "Aged Brie" :sell-in 2 :quality 10}
   {:name "Elixir of the Mongoose" :sell-in 5 :quality 7}
   {:name "Sulfuras, Hand of Ragnaros" :sell-in 0 :quality 80}
   {:name "Backstage passes to a TAFKAL80ETC concert" :sell-in 15 :quality 20}
   {:name "Conjured Mana Cake" :sell-in 3 :quality 6}])

(defmulti update-quality-mm :name)

(defmethod update-quality-mm "Aged Brie"
  [{:keys [name sell-in quality] :as item}]
  (-> item
      (update :sell-in dec)
      (update :quality #(min (if (<= sell-in 0)
                               (+ 2 %)
                               (+ 1 %))
                             50))))

(defmethod update-quality-mm "Backstage passes to a TAFKAL80ETC concert"
  [{:keys [name sell-in quality] :as item}]
  (-> item
      (update :sell-in dec)
      (update :quality #(min (condp >= sell-in
                               0  0
                               5  (+ 3 %)
                               10 (+ 2 %)
                               (+ 1 %))
                             50))))

(defmethod update-quality-mm "+5 Dexterity Vest"
  [{:keys [name sell-in quality] :as item}]
  (-> item
      (update :sell-in dec)
      (update :quality #(max (if (<= sell-in 0)
                               (- % 2)
                               (- % 1))
                             0))))

(defmethod update-quality-mm "Elixir of the Mongoose"
  [{:keys [name sell-in quality] :as item}]
  (-> item
      (update :sell-in dec)
      (update :quality #(max (if (<= sell-in 0)
                               (- % 2)
                               (- % 1))
                             0))))

(defmethod update-quality-mm "Sulfuras, Hand of Ragnaros"
  [{:keys [name sell-in quality] :as item}]
  item)

(defmethod update-quality-mm "Conjured Mana Cake"
  [{:keys [name sell-in quality] :as item}]
  (-> item
      (update :sell-in dec)
      (update :quality #(max (if (<= sell-in 0)
                               (- % 4)
                               (- % 2))
                             0))))

(defn update-quality
  "Your implementatinon here.  Method body can be changed,but not
  the signature.

  Each day reduce both quality and sell-in by the amt required for
  the relevant item."
  [item]
  (update-quality-mm item))

(defn days [item n]
  (nth (iterate update-quality item) n))

(defn elapsed-days [n]
  (mapv #(days % n)
        items))
